### ISSUE TITLE
Dangerous behaviour of TimeoutAttribute on NET Core and later

### ISSUE THREAD

**Issue by manfred-brands** [nunit/nunit#4021]
Due to the [removal of `Thread.Abort` in later versions of .NET](https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/5.0/thread-abort-obsolete), the `TimeoutAttribute` no longer stops the long running test.

This means the test keeps running in the background, despite that the nunit runner moves to the next test.
As an example, see the following testfixture:

```
internal class TimeOut
{
    private string _logFilePath;

    [OneTimeSetUp]
    public void OneTimeSetup()
    {
        _logFilePath = Path.Combine(TestContext.CurrentContext.TestDirectory, "Timeout.txt");
        File.Delete(_logFilePath);
    }

    private void WriteLine(string text)
    {
        File.AppendAllText(_logFilePath, text + Environment.NewLine);
    }

    [TearDown]
    public void Teardown()
    {
        WriteLine("Teardown, result: " + TestContext.CurrentContext.Result.Outcome);
    }

    [Test]
    [Order(1)]
    [Timeout(110)]
    public void WithTimeout()
    {
        LongTest("WithTimeout");
    }

    [Test]
    [Order(2)]
    public void NoTimeout()
    {
        LongTest("NoTimeout");
    }

    private void LongTest(string testName)
    {
        try
        {
            for (int i = 0; i < 5; i++)
            {
                WriteLine($"{testName}: Iteration {i}");
                Thread.Sleep(100);
            }
            Assert.Pass("Success");
        }
        catch (Exception ex)
        {
            WriteLine(ex.Message);
        }
        finally
        {
            WriteLine("Task finished");
        }
    }
}
```

When running this on NET48, the `net48/Timeout.txt' file contains:

```
WithTimeout: Iteration 0
WithTimeout: Iteration 1
Thread was being aborted.
Task finished
Teardown, result: Failed:Cancelled
NoTimeout: Iteration 0
NoTimeout: Iteration 1
NoTimeout: Iteration 2
NoTimeout: Iteration 3
NoTimeout: Iteration 4
Task finished
Teardown, result: Passed
```

This shows the 1st test being aborted, and its `Teardown` method running, before the 2nd test starts.

When running on .NET5, the `net5.0/Timeout.txt` contains:

```
WithTimeout: Iteration 0
WithTimeout: Iteration 1
NoTimeout: Iteration 0
WithTimeout: Iteration 2
NoTimeout: Iteration 1
WithTimeout: Iteration 3
NoTimeout: Iteration 2
WithTimeout: Iteration 4
NoTimeout: Iteration 3
Task finished
NoTimeout: Iteration 4
Task finished
Teardown, result: Passed
```

This shows that the `WithTimeout` test runs to the end but with no `Teardown` running.
At the same time, the 2nd test is running in parallel, starting after the timeout of the 1st test.

There are multiple issues:
1. `TearDown` is not run when the test 'times out'
2. The test keeps running, which on tests with shared instance fields could be problematic, possibly interfering with other tests.

If I add a `_counter` field, set to `0` in `SetUp`, increment it in the for-loop and `Assert.That(_counter, Is.EqualTo(5));`, in the actual test, the net5.0 run fails with:
```
 Failed NoTimeout [507 ms]
  Error Message:
     Expected: 5
  But was:  8

  Stack Trace:
     at Tests.TimeOut.NoTimeout() in C:\Users\m.brands\source\Tests\TimeOut.cs:line 53
```

The question is what can be done about it.
1. Mark the `TimeoutAttribute` as obsolete except for the build targeting net framework?
2. Add an analyzer rule in nunit.analyzers warning about the drawbacks on non-framework builds.
3. Can we fix the behaviour to what it was before. Not likely.
4. Add a `CancellationToken` to `TestContext`, set to either `CancellationToken.None` if no timeout set, or a token that will be cancelled after the time out, see below. This approach will require updating end-users tests to 'the new way'.
```
using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(110);
await LongTest("WithTimeout", cancellationTokenSource.Token);
Assert.That(_counter, Is.EqualTo(5));
```

**Comment by Dreamescaper** [comment]
Another option - change TimeoutAttribute behavior to terminate the whole run if it exceeds. Might be useful simply to prevent the test run from hanging.

**Comment by stevenaw** [comment]
Thanks for catching this @manfred-brands .
I like the idea of exposing a CancellationToken on the test context myself. It seems like the least-intrusive opt-in approach here which is non-breaking and would work equally well to help address this in both 3.x and 4.0.
@nunit/framework-team what do you think?

**Comment by jnm2** [comment]
I like the CancellationToken idea. It would also be idiomatic .NET (and cool) if it could be injected as a test method parameter, but that could be done later.

**Comment by manfred-brands** [comment]
I will see if I can create a PR with the CancellationToken as a proof of concept.

**Comment by Dreamescaper** [comment]
I feel like this CancellationToken approach won't be used in 99.9% of cases.
OTOH, the current approach might be not that bad. If I simply want to make sure that my test run won't hang, abandoning the task is not the worst idea.
But we need to update our documentation to mention that, maybe even mention in test failure message that other test results might be affected.

**Comment by manfred-brands** [comment]
> I feel like this CancellationToken approach won't be used in 99.9% of cases.

Neither is the `TimeoutAttribute` in the first place. We have 13 usages in 126k test.
I only see one usage for `TimeoutAttribute`: For integration tests that actually communicate with external services.
Newer code should already use a `CancellationToken` for this kind of code and could benefit from this.

>If I simply want to make sure that my test run won't hang, abandoning the task is not the worst idea.

The problem is that `TearDown` actions are not executed, potentially leaving files/directories/databases behind.
Other resources are less of an issue as they would be released on process exit. As timeout is a likely a random failure, "fixed" on the next test run these are never addressed.

The hanging of one test could also affect the next if the test hangs on a lock.

**Comment by bormm** [comment]
I think there is a common understanding what a Timeout mean in general. (For me) Its the time a operation should not exceed to run. 

But there it ends already. So what happens on a timeout is not a common understanding.
Its also limited by the architecture and implementation of the test framework, the used runtime environment and the stuff that is tested of course.

In NUnit we also have MaxTimeout, which is by naming something redundant, because a timeout is a maximum time already. But by implementation there is of course a difference between MaxTimeout (just failure if exceeded) where Timeout terminated in the past the test but now sometimes doesn't, without any chance to understand this without getting deeper.

So my ideas: As there is no common action possible and with MaxTimeout() everyone can see there was already a different expectation in the past, this can maybe merged into one thing: make it a mandatory option to choose.

What was MaxTimeout in the past could be something like [Timeout(Timespan=1000, Action=Report)]
And the other options could be
 - Terminate, which is then unavailable if using .Net Core
 - Cancel: using a CancellationToken
 - ReportAndContinueInBackground: As the current .Net Core behavior

For the CancellationToken I think thats maybe a good idea for very new code that is very good designed. But for the grown code I work on here that is created in many years, there is not a single CancellationToken in the code at all. Would be easier to throw all away than adding CancellationToken.

For the TearDown issue that @manfred-brands originally pointed here, I see this as a bug in NUnit. In case terminate is not possible, it can just not continue with the other tests before being able to call TearDown.
If NUnit detects this situation, like if choose ReportAndContinueInBackground + TearDown exists, it should terminate the whole run with a appropriate error message.

I also wonder if somehow the --blame-hang in dotnet test can be triggered here also.

**Comment by manfred-brands** [comment]
@bormm The problem with the `TearDown` is that it might fail. We have `TearDown` methods that remove temporary directories/files. This fails (at least under Windows), if something has the file still open. On the other hand, the `TearDown` method might `Dispose` items which might cause the hanging test to finish. I will see if I can get the `TearDown` method to be called on a `Timeout`.

**Comment by bormm** [comment]
@manfred-brands I totally understood your problem and I didn't criticize you at all, quite the opposite.

Until there is some solution in NUnit, I think nobody should use Timeout anymore but MaxTimeout only. That would take sure there is a validation and error if a test took longer than expected.
To catch cases where test hang, which is typically a bug, then it will hang typically also forever. This means running all tests with --blame-hang and specifying a timeout that is larger than any expected test run, seems the best solution currently.

**Comment by eatdrinksleepcode** [comment]
> I only see one usage for TimeoutAttribute: For integration tests that actually communicate with external services.
Newer code should already use a CancellationToken for this kind of code and could benefit from this.

This is _not_ the only use for `TimeoutAttribute`. I want to have a per-test global timeout configured for my unit tests, so that if I end up with a bug that causes an infinite loop or other runaway behavior, the test is terminated and I am informed of the problem, instead of just getting a hung process with no useful information as to the cause. `CancellationToken` is _not_ useful for this in the general case.

**Comment by eatdrinksleepcode** [comment]
A new API was recently implemented as a replacement for `Thread.Abort`: [`ControlledExecution`](https://github.com/dotnet/runtime/pull/71661). Only available in .NET 7+, but it might be capable of re-enabling the core premise of `TimeoutAttribute`.

**Comment by manfred-brands** [comment]
Nice find. 
It looks like they re-introduced `Thread.Abort` by another name and then immediately told people not to use it for the same reasons as before: [SYSLIB0046](https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0046)
Note the comment: _While this method is new for .NET 7, it's also marked as obsolete to discourage you from using it._

With the known caveats, I think it can be used for NUnit tests as long as we ensure `TearDown` methods are executed on timed-out tests. It is then up to the user to deal with cleaning up after timeout abortions.

**Comment by manfred-brands** [comment]
`TimeoutAttribute` is marked `Obsolete` for .NET Core builds.
Test Code should be upgraded to use the `CancelAfterAttribute` instead.

**Comment by herebebeasties** [comment]
> `TimeoutAttribute` is marked `Obsolete` for .NET Core builds. Test Code should be upgraded to use the `CancelAfterAttribute` instead.

That's all very well, but you can't set it at an assembly level, only on methods/classes.
The previous `TimeoutAttribute` could be set on a whole assembly, which covered the use case of preventing infinite test hangs in CI, etc. What's the recommended new equivalent?

**Comment by manfred-brands** [comment]
@herebebeasties for a global timeout I suggest to use 

`dotnet test --blame-hang-timeout <TIMESPAN>`

See [dotnet cli](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-test#options)
